File: language-testing/udp_echo_client/udp_echo_client.pl
#! /usr/bin/env perl

use strict;
use warnings;
use IO::Socket;

sub send_receive_data {
  my $sock = new IO::Socket::INET (
    PeerAddr => "$_[0]:$_[1]",
    Proto => "udp",
  ) or die "Socket error: $!\n";

  print $sock $_[2];
  my $data;
  $sock->recv($data, 1024);
  print $data;
  close $sock;
}

my $usage = "Usage: $0 hostname port";
die "$usage\nError: No hostname provided!" unless $ARGV[0];
die "$usage\nError: No port provided!" unless $ARGV[1];
die "$usage\nError: Port contains non-digits!" unless $ARGV[1] =~ /^\d+$/;
&send_receive_data($ARGV[0], $ARGV[1], <STDIN>);

File: language-testing/udp_echo_client/udp_echo_client.rb
#! /usr/bin/env ruby

require 'socket'

def send_receive_data hostname, port, content
  UDPSocket.open do |sock|
    sock.send content, 0, hostname, port
    data, _ = sock.recvfrom(1024)
    puts data
  end
end

def die msg
  $stderr.puts "Usage: #{$0} hostname port\nError: #{msg}" 
  exit 1
end

if __FILE__ == $0
  die "No hostname given!" unless ARGV.length > 0
  die "No port given!" unless ARGV.length > 1
  die "Port contains non-numbers!" unless /^\d+$/ =~ ARGV[1]
  die "Too many arguments given!" unless ARGV.length == 2
  send_receive_data ARGV[0], ARGV[1].to_i, $stdin.read
end
